Sources/Basics/Archiver+Tar.swift

```diff
@@ -0,0 +1,147 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift open source project
+//
+// Copyright (c) 2014-2023 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See http://swift.org/LICENSE.txt for license information
+// See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import class Dispatch.DispatchQueue
+import struct Dispatch.DispatchTime
+import struct TSCBasic.AbsolutePath
+import protocol TSCBasic.FileSystem
+import struct TSCBasic.FileSystemError
+import class TSCBasic.Process
+
+/// An `Archiver` that handles Tar archives using the command-line `tar` tool.
+public final class TarArchiver: Archiver {
+    public var supportedExtensions: Set<String> { ["tar", "tar.gz"] }
+
+    /// The file-system implementation used for various file-system operations and checks.
+    private let fileSystem: FileSystem
+
+    /// Helper for cancelling in-flight requests
+    private let cancellator: Cancellator
+
+    /// The underlying command
+    private let tarCommand: String
+
+    /// Creates a `TarArchiver`.
+    ///
+    /// - Parameters:
+    ///   - fileSystem: The file system to used by the `TarArchiver`.
+    ///   - cancellator: Cancellation handler
+    public init(fileSystem: FileSystem, cancellator: Cancellator? = .none) {
+        self.fileSystem = fileSystem
+        self.cancellator = cancellator ?? Cancellator(observabilityScope: .none)
+
+        #if os(Windows)
+        self.tarCommand = "tar.exe"
+        #else
+        self.tarCommand = "tar"
+        #endif
+    }
+
+    public func extract(
+        from archivePath: AbsolutePath,
+        to destinationPath: AbsolutePath,
+        completion: @escaping (Result<Void, Error>) -> Void
+    ) {
+        do {
+            guard self.fileSystem.exists(archivePath) else {
+                throw FileSystemError(.noEntry, archivePath)
+            }
+
+            guard self.fileSystem.isDirectory(destinationPath) else {
+                throw FileSystemError(.notDirectory, destinationPath)
+            }
+
+            let process = TSCBasic.Process(
+                arguments: [self.tarCommand, "xzf", archivePath.pathString, "-C", destinationPath.pathString]
+            )
+
+            guard let registrationKey = self.cancellator.register(process) else {
+                throw StringError("cancellation")
```

Comment:
nicer error message?

Reply:
This duplicates the error message from `ZipArchiver`, would it be ok to update the error message there as well in the same PR?

Reply:
please

Reply:
I've added a new `static func` on `CancellationError` that populates its `description` with a more detailed message, now used across both `TarArchiver` and `ZipArchiver`.